core: improve clist

kaspar030 opened the pull request:
After working with redis for a while, I now find it's naming for list operations way more intuitive than what we had before, so I adapted our clist implementation accordingly.

While at it, I added some functions and (IMHO) improved the docs quite a bit.

*edit*

With the improvements from this PR, clist becomes a generally usable data structure that can be used wherever a list, stack (lifo) or queue (fifo) is needed, with O(1) operations for adding objects at head or tail and removing objects from head, in addition to O(n) operations for remove from tail, remove any object and find any object.

gebart commented:
Could you add an implementation for `clist_print` as well while you're working on clist?

gebart commented:
I like the naming changes, it makes the API clearer.
I have tested some of the changes in the unit tests for VFS in #5616 (`clist_find`, `clist_remove`, `clist_rpush`)

gebart commented on core/include/clist.h:
```diff
@@ -14,20 +14,64 @@
+ *
+ * Implementation details:
  *
  * Each list is represented as a "clist_node_t". It's only member, the "next"
```
"It's" -> "Its"

kaspar030 replied:
addressed

gebart replied:
did you forget to git push?

gebart commented on core/include/clist.h:
```diff
@@ -14,20 +14,64 @@
 #ifndef CLIST_H
 #define CLIST_H
 
+#include <stdio.h>
```
not used anywhere as far as I can see?

kaspar030 replied:
yup, addressed.

gebart commented:
Can you improve the unit test coverage? `clist_lpush` and `clist_rpop` are not tested currently

gebart commented:
(https://github.com/RIOT-OS/RIOT/pull/5616#issuecomment-233168950)
@kaspar030 wrote:
>you not feeling confident about it's usage makes me wonder if the API is intuitive like this?

I don't think there's anything wrong with the API in this PR, rather the documentation is missing some pieces. The missing piece for me was the fact that the list is represented by a `clist_node_t ` item which is outside of the list circle but pointing in (`_vfs_mounts_list` in #5616).

Another thing is the iteration of lists. The example could be made more concrete, I'll try to come up with some improvements and post back here.
The name of the function `clist_advance` made me think that I should use that for iteration. Maybe the example code should use that?

I think this does the same thing but with `clist_advance` instead of manually advancing, with the downside that the `container_of` macro needs to be passed the `node->next` pointer instead of just `node` (untested code)
```c
/* outside of function */
struct {
    int value1;
    char str[8];
    clist_node_t list_entry;
} myitem_t;

/* List handle */
clist_node_t mylist;

/* In function where we need to iterate over the list once: */

/* Make a copy of the list handle to use for iteration */
clist_node_t node = mylist;

if (node->next == NULL) {
    /* list empty */
    /* do some special handling */
    return;
}
do {
    /* advance the iteration handle to point at the next item */
    clist_advance(node);
    /* find a myitem_t pointer given a pointer to item->list_entry */
    myitem_t *item = container_of(node->next, myitem_t, list_entry);
    /* do something with *item */
    /* example: */
    if (strcmp(item->str, "hej") == 0) {
        return item;
    }
    /* ... */
    /* check if we have gone round the list to the last item */
} while (node.next != _vfs_mounts_list.next);
```

gebart commented:
Let's see what Murdock says.

kaspar030 commented:
- updated unittests
- fixed headers

I've added a "clist_foreach()" function that can be used for traversal using a function pointer (and added an example on how to use it to print clist entries).

How about renaming "clist_advance()" to "clist_rotate()", "clist_rotate_clockwise()" or "clist_rotate_right()"?

Would it make sense to typedef "clist_t" to "clist_node_t" and then use the former as list reference object?

miri64 commented on core/include/clist.h:
```diff
@@ -40,23 +93,19 @@
- * @brief inserts *new_node* into *list*
+ * @brief Appends *new_node* at the end of *list
+ *
+ * @note Complexity: O(1)
```
Matter of taste, but surrounding `$` to change to LaTeX parsing?

miri64 replied:
(to be in line with other, more complex mathematical terms)

gebart replied:
if you want to use LaTeX parsing for the complexity you might as well go all in and write it `$\mathcal{O}(n)$`

see http://texblog.org/2014/06/24/big-o-and-related-notations-in-latex/ for an example

kaspar030 replied:
If you don't insist, I would keep it ascii, as some people do look at the code to read the documentation, and in that case ```O(n)``` is a lot more pleasant to the eyes than ```thcal{O}(n)$```.

miri64 replied:
$O(1)$

gebart commented:
After this is merged, does it make sense to deprecate the other `list.h` and update the users to use `clist.h` instead?
`list_remove_head` should be replaceable by `clist_lpop`, if I am not mistaken.
Current users of `list_remove_head` are `mutex.c` and `msg.c`.
`list_add` is not used anywhere in the main repository, as far as I can see.

gebart commented:
Why is not `clist_foreach` inlined when everything else is?
The `clist_foreach` function does not look very much more complex than anything else in the `clist.h` header and it would likely give the compiler an opportunity for some optimizations for the case when the function pointer is a compile time constant.

miri64 commented:
> After this is merged, does it make sense to deprecate the other list.h and update the users to use clist.h instead?

So back to how it used to be ^^?

gebart commented:
I don't think clist_rotate_right/clockwise are more speaking than clist_advance. In fact, I find it hard to understand which part is moving clockwise, depending on how you picture the list.

If you picture the circular list items as the face of a clock and the list handle is the hand of the clock, then it makes sense to call it rotate clockwise, but on the other hand, if you picture the list handle as fixed and the circular list as the moving part (wheel of fortune style), then rotate clockwise is the opposite operation.

I have never worked with Redis but I looked up the manual after I read your description on this PR. There is an operation called RPOPLPUSH which does the same thing, except it takes two arguments, see http://redis.io/commands/rpoplpush. Maybe that is something to consider for the `clist_advance` functionality?

gebart commented:
@miri64 you are right (https://github.com/RIOT-OS/RIOT/pull/4557 introduced the other list.h). I see no reason for list.h to live on after this is merged.

kaspar030 commented:
> After this is merged, does it make sense to deprecate the other list.h and update the users to use clist.h instead?

I thought about it, but then again, clist has (small) overhead over list.h, and it's main users are msg and mutex, where that kind of overhead is actually visible.

kaspar030 commented:
> RPOPLPUSH which does the same thing, except it takes two arguments

Hm. That operation is supposed to be used for reliable queueing (rpop from one list, lpush to another, atomically), and that it can be used on the same list is a side effect. But then it does what clist_advance() does (or better, lpoprpush would do). I think I like it better than _advance(), but I'm a little afraid that people forget this is still a circular list. :)

kaspar030 commented:
> So back to how it used to be ^^?

mutex and msg used priority_queue before.

kaspar030 commented:
- rebased
- renamed ```clist_advance()``` -> ```clist_lpoprpush()```
- added ```clist_lpeek()``` and ```clist_rpeek()```

gebart commented:
@kaspar030 
>*edit* I'll do some more benchmarking, but let's keep list.h for now (this PR), ok?

Yes, let's leave the list.h change for a separate PR.

kaspar030 commented:
I think this should be mergable, so I'll squash. @gebart @miri64 do you feel like ACK today? ;)

gebart commented:
Please add a minimal unit test for `clist_foreach`.
ACK from me once that is done. Needs a second ACK as well

miri64 commented on core/include/clist.h:
```diff
@@ -98,23 +171,175 @@ static inline clist_node_t *clist_remove_head(clist_node_t *list)
+ * @returns         node if found and removed
+ * @returns         NULL if node is not a list member
+ */
+static inline clist_node_t *clist_remove(clist_node_t *list, clist_node_t *node)
```
I think this function is big enough to be a non-inline function.

kaspar030 replied:
Some of them might be. I kept them static inline as they're used by the scheduler, where every instruction counts. I'll do an evaluation of the benefits/downsides in a folo.

miri64 replied:
kk.

miri64 commented on tests/unittests/tests-core/tests-core-clist.c:
```diff
@@ -22,84 +22,198 @@ static list_node_t test_clist;
+
+static void test_clist_lpush(void)
+{
+    set_up();
```
set_up() is called automatically before this function is called so why do you call it?

miri64 commented on core/include/clist.h:
```diff
@@ -98,23 +171,175 @@ static inline clist_node_t *clist_remove_head(clist_node_t *list)
+ * @param[in]       list        List to traverse.
+ * @param[in]       func        Function to call for each member.
+ */
+static inline void clist_foreach(clist_node_t *list, int(*func)(clist_node_t *))
```
This one too, maybe.

gebart replied:
I wrote in an earlier comment that this function should be inlined to allow the compiler to optimize for the case where `func` is compile time constant, which is likely the most common case.
(the comment was deleted by github when the file no longer exists)

miri64 replied:
Okay, sorry for the noise then.

kaspar030 commented:
- added clist_foreach() unittest, immediately squashed

kaspar030 commented:
hmm, murdock is not setting the new status.

miri64 commented on tests/unittests/tests-core/tests-core-clist.c:
```diff
@@ -70,36 +173,107 @@ static void test_clist_remove_two(void)
+
+static void test_clist_foreach(void)
+{
+    set_up();
```
- set_up() ;-)

kaspar030 replied:
clist needs to be clear before the test starts, but this test doesn't call one of the _add_n() functions.

miri64 replied:
But `set_up()` is already run by the test system before the call to this function to ensure that.

miri64 commented on tests/unittests/tests-core/tests-core-clist.c:
```diff
@@ -22,46 +22,149 @@ static list_node_t test_clist;
 
 static void test_clist_add_two(void)
 {
+    set_up();
```
\- set_up()

kaspar030 replied:
nope, this function is called from the other tests, and they expect a defined state.

miri64 replied:
Ah, kk. Can you please mention this in a comment?

miri64 commented on tests/unittests/tests-core/tests-core-clist.c:
```diff
@@ -22,46 +22,149 @@ static list_node_t test_clist;
-static void test_clist_add_one(void)
+static void test_clist_rpush(void)
 {
+    set_up();
```
\- set_up()

kaspar030 commented:
@miri64 Didn't know set_up() was called.

- removed redundant set_up() calls
- immediately squashed

kaspar030 commented:
Again the status doesn't update. The build was successful (https://ci.riot-labs.de/RIOT-OS/RIOT/5629/a3547963dc4afe001dc37599304c60f784fd2b3b/output.html).

kaspar030 commented:
Now it's green. :)

miri64 commented:
ACK and go.

basilfx commented:
@kaspar030 I get a build error due to missing return in this method (using Clang): https://github.com/RIOT-OS/RIOT/blob/master/core/include/clist.h#L195

The `return NULL;` is missing.

kaspar030 commented:
@basilfx yup, I PR'ed a fix.
